static/js/containers/ProfilePage.js

```diff
@@ -42,6 +42,7 @@ class ProfilePage extends React.Component {
     return dispatch(validateProfile(profile, requiredFields, messages)).then(() => {
       dispatch(saveProfile(SETTINGS.username, profile)).then(() => {
         dispatch(clearProfileEdit());
+        return Promise.resolve();
```

Comment:
the race condition we discovered is still occurring with this change. I posted a comment on the pr showing the correct way to change this function:

```js
  saveProfile(isEdit, profile, requiredFields, messages) {
    const { dispatch } = this.props;
    if (!isEdit) {
      // Validation errors will only show up if we start the edit
      dispatch(startProfileEdit());
    }
    return dispatch(validateProfile(profile, requiredFields, messages)).then(() => {
      return dispatch(saveProfile(SETTINGS.username, profile)).then(() => {
        dispatch(clearProfileEdit());
        return Promise.resolve();
      });
    });
  }
```

basically, each of the chained arrow functions needs to return a promise, this is what gives us guarantees of the statements being executed synchronously.

you're change returns `Promise.resolve()` at the end, but because the first arrow function (opened on line 42) doesn't return a promise we aren't actually going to be chaining them together. if you change this function to what I have above we should get the behavior that we want, which is the actions happening strictly in this order:

```
validateProfile -> saveProfile -> clearProfileEdit -> return Promise.resolve()
```

then, with the guarantees around strict order of execution that this gives us, we can ensure that we won't trigger a route transition (to `/dashboard`) before the `PATCH` request comes back. The problem before was this line (in the `saveAndContinue` function, defined in `util/profile_edit.js`:

```js
  saveprofile( profile, fields, validationmessages).then(() => {
    this.context.router.push(next);
  });
```

before we weren't chaining promises, so we didn't actually have any kind of guarantee that our callback (`() => this.context.router.push(next);`) would execute after all of the actions in `saveProfile` had finished dispatching. With the change above we do have such a guarantee - so we can do the route transition ( `this.context.router....`) safely, knowing that we're all done with saving the profile, and then we can safely dispatch the `startProfileEdit` and `updateValidation` actions when we call the `requireCompleteProfile` function.

Does this all make sense?